crates/bevy_ecs/src/template.rs

```diff
@@ -0,0 +1,190 @@
+//! Functionality that relates to the [`Template`] trait.
+
+pub use bevy_ecs_macros::GetTemplate;
+
+use crate::{
+    bundle::Bundle,
+    entity::{Entity, EntityPath},
+    error::{BevyError, Result},
+    world::EntityWorldMut,
+};
+use alloc::{boxed::Box, vec, vec::Vec};
+use bevy_platform::collections::hash_map::Entry;
+use bevy_utils::TypeIdMap;
+use core::any::{Any, TypeId};
+use downcast_rs::{impl_downcast, Downcast};
+use variadics_please::all_tuples;
+
+/// A [`Template`] is something that, given a spawn context (target [`Entity`], [`World`](crate::world::World), etc), can produce a [`Template::Output`].
+pub trait Template {
+    /// The type of value produced by this [`Template`].
+    type Output;
+
+    /// Uses this template and the given `entity` context to produce a [`Template::Output`].
+    fn build(&mut self, entity: &mut EntityWorldMut) -> Result<Self::Output>;
+
+    /// This is used to register information about the template, such as dependencies that should be loaded before it is instantiated.
+    #[inline]
+    fn register_data(&self, _data: &mut TemplateData) {}
+}
+
+/// [`GetTemplate`] is implemented for types that can be produced by a specific, canonical [`Template`]. This creates a way to correlate to the [`Template`] using the
+/// desired template output type. This is used by Bevy's scene system.
+pub trait GetTemplate: Sized {
```

Comment:
What role does `FromWorld` play in a post-`GetTemplate` Bevy? I would prefer if it was completely replaced: I don't like the arbitrary mutable side effects.

But this is just a gut feeling now: there may be barriers I'm missing.

Reply:
We *should* be able to instantiate resources via templates, especially in a resources-as-components Bevy. I agree though, it doesn't look like it's currently implemented.

Reply:
I had a similar question about `FromWorld`. Can we completely replace `FromWorld` with `GetTemplate`?

If so, should we rename `GetTemplate` to `FromWorld`? I find that name much more self-explanatory. If there's a canonical way to create a type from the world, parameterized by some `Template` type `T`, it should implement `FromWorld<Template=T>`